cumulus_library/statistics/psm.py

```diff
@@ -0,0 +1,247 @@
+# Module for generating Propensity Score matching cohorts
+
+import numpy as np
+import pandas
+import toml
+
+from psmpy import PsmPy
+
+
+import json
+from pathlib import PosixPath
+from dataclasses import dataclass
+
+from cumulus_library.cli import StudyBuilder
+from cumulus_library.databases import DatabaseCursor
+from cumulus_library.base_table_builder import BaseTableBuilder
+from cumulus_library.template_sql.templates import (
+    get_ctas_query_from_df,
+    get_drop_view_table,
+)
+from cumulus_library.template_sql.statistics.psm_templates import (
+    get_distinct_ids,
+    get_create_covariate_table,
+)
+
+
+@dataclass
+class PsmConfig:
+    """Provides expected values for PSM execution
+
+    These values should be read in from a toml configuration file.
+    See tests/test_data/psm/psm_config.toml for an example with details about
+    the expected values for these fields.
+    """
+
+    classification_json: str
+    pos_source_table: str
+    neg_source_table: str
+    target_table: str
+    primary_ref: str
+    count_ref: str
+    count_table: str
+    dependent_variable: str
+    pos_sample_size: int
+    neg_sample_size: int
+    join_cols_by_table: dict[str, dict]
+    seed: int
+
+
+class PsmBuilder(BaseTableBuilder):
+    """TableBuilder for creating PSM tables"""
+
+    display_text = "Building PSM tables..."
+
+    def __init__(self, toml_config_path: str):
+        """Loads PSM job details from a PSM configuration file"""
+        with open(toml_config_path, encoding="UTF-8") as file:
+            toml_config = toml.load(file)
+        self.config = PsmConfig(
+            classification_json=f"{PosixPath(toml_config_path).parent}/{toml_config['classification_json']}",
+            pos_source_table=toml_config["pos_source_table"],
+            neg_source_table=toml_config["neg_source_table"],
+            target_table=toml_config["target_table"],
+            primary_ref=toml_config["primary_ref"],
+            dependent_variable=toml_config["dependent_variable"],
+            pos_sample_size=toml_config["pos_sample_size"],
+            neg_sample_size=toml_config["neg_sample_size"],
+            join_cols_by_table=toml_config.get("join_cols_by_table", {}),
+            count_ref=toml_config.get("count_ref", None),
+            count_table=toml_config.get("count_table", None),
+            seed=toml_config.get("seed", 123),
+        )
+        super().__init__()
+
+    def _get_symptoms_dict(self, path: str) -> dict:
+        """convenience function for loading symptoms dictionaries from a json file"""
+        with open(path) as f:
```

Comment:
I'll be your pylint call and say you're missing an `encoding` arg.

Reply:
Ah maybe I'm thinking of the pep8 checker `pycodestyle` then. Or we have different pylint configs. I have fixed many a missing `encoding` arg.